project-koku/koku
# [COST-4028] - data migration

## Jira Ticket

[COST-4028](https://issues.redhat.com/browse/COST-4028)

## Description

This change will properly populate the `reporting_ocp_pod_summary_by_node_p` and move any data that may be in the `_default` partition into the correct tables.

We ran a script in production (step 8 below) which did not put the data into the correct partitions. The migration in this PR will call the `PartitionHandlerMixin()._handle_partitions` which will move data from the default partition into the correct monthly partitions.

After this migration is run in production, we can convert it into a no-op.

## Testing

1. apply the following diff:
```
--- a/dev/scripts/load_test_customer_data.sh
+++ b/dev/scripts/load_test_customer_data.sh
@@ -79,7 +79,7 @@ export S3_BUCKET_NAME="ocp-ingress"
 
 log-info "Calculating dates..."
 if [[ $OS = "Darwin" ]]; then
-    START_DATE=${2:-$(date -v '-1m' +'%Y-%m-01')}
+    START_DATE=${2:-$(date -v '-4m' +'%Y-%m-01')}
 else
     START_DATE=${2:-$(date --date='last month' '+%Y-%m-01')}
 fi
diff --git a/koku/reporting/provider/ocp/models.py b/koku/reporting/provider/ocp/models.py
index 20d0caec..579029f7 100644
--- a/koku/reporting/provider/ocp/models.py
+++ b/koku/reporting/provider/ocp/models.py
@@ -44,7 +44,7 @@ UI_SUMMARY_TABLES_MARKUP_SUBSET = (
 UI_SUMMARY_TABLES = (
     *UI_SUMMARY_TABLES_MARKUP_SUBSET,
     "reporting_ocp_pod_summary_p",
-    "reporting_ocp_pod_summary_by_node_p",
+    # "reporting_ocp_pod_summary_by_node_p",
     "reporting_ocp_pod_summary_by_project_p",
     "reporting_ocp_volume_summary_p",
     "reporting_ocp_volume_summary_by_project_p",
```
2. set `RETAIN_NUM_MONTHS=6`
3. start with a fresh db `make delete-db (requires a full restart of everything)` but do NOT apply this migration yet
5. `$ make create-test-customer; make load-test-customer-data test_source=onprem`
6. to ensure there is data back to august, do:
http://localhost:5042/api/cost-management/v1/report_data/?schema=org1234567&start_date=2023-07-01&provider_type=OCP&provider_uuid={provider-uuid-of-your-OCP-provider}
7. let the data summarize - this will take some time, so be patient.
8. See data for all months here:
http://localhost:8000/api/cost-management/v1/reports/openshift/compute/?filter[offset]=0&end_date=2023-11-06&start_date=2023-08-01
9. run this script in a django shell (`make shell`):
```
from api.models import Provider
from masu.database.ocp_report_db_accessor import OCPReportDBAccessor

p = Provider.objects.filter(type='OCP', active=True, paused=False)

tables = ['reporting_ocp_pod_summary_by_node_p']
for prov in p:
    try:
        with OCPReportDBAccessor(prov.customer.schema_name) as acc:
            acc.populate_ui_summary_tables('2023-08-01', '2023-10-01', str(prov.uuid), tables=tables)
    except Exception as err:
        print(err)
```
10. In the db, run some commands and see data in the `_default` partition and see that the real partitions are missing:
```
postgres=# select count(*) from org1234567.reporting_ocp_pod_summary_by_node_p;
 count 
-------
  2790
(1 row)

postgres=# select count(*) from org1234567.reporting_ocp_pod_summary_by_node_p_default; <<<< this data is in the wrong table
 count 
-------
  2790
(1 row)

postgres=# select count(*) from org1234567.reporting_ocp_pod_summary_by_node_p_2023_10;
ERROR:  relation "org1234567.reporting_ocp_pod_summary_by_node_p_2023_11" does not exist
LINE 1: select count(*) from org1234567.reporting_ocp_pod_summary_by...
                             ^
postgres=# select count(*) from org1234567.reporting_ocp_pod_summary_by_node_p_2023_09;
ERROR:  relation "org1234567.reporting_ocp_pod_summary_by_node_p_2023_10" does not exist
LINE 1: select count(*) from org1234567.reporting_ocp_pod_summary_by...
```
11. Run the migration in this PR: `make run-migrations`
12. See that there is now no data in the `_default` table and the data is now in the actual partitions:
```
postgres=# select count(*) from org1234567.reporting_ocp_pod_summary_by_node_p_default;
-[ RECORD 1 ]
count | 0

postgres=# select count(*) from org1234567.reporting_ocp_pod_summary_by_node_p_2023_10;
-[ RECORD 1 ]
count | 1350

postgres=# select count(*) from org1234567.reporting_ocp_pod_summary_by_node_p_2023_09;
-[ RECORD 1 ]
count | 1350

postgres=# select count(*) from org1234567.reporting_ocp_pod_summary_by_node_p_2023_08;
-[ RECORD 1 ]
count | 1395

```
## Notes

...
